packages/gatsby-cli/src/reporter/reporter.ts

```diff
@@ -0,0 +1,267 @@
+import { stripIndent } from "common-tags"
+import chalk from "chalk"
+import { trackError } from "gatsby-telemetry"
+import { globalTracer, Span } from "opentracing"
+
+import reporterActions from "./redux/actions"
+import { LogLevels, ActivityStatuses } from "./constants"
+import { getErrorFormatter } from "./errors"
+import constructError from "../structured-errors/construct-error"
+import { prematureEnd } from "./short-circut"
+import { IStructuredError } from "../structured-errors/types"
+import { createTimerReporter, TimerReporter } from "./reporter-timer"
+import { createPhantomReporter, PhantomReporter } from "./reporter-phantom"
+import { createProgressReporter, ProgressReporter } from "./reporter-progress"
+
+const errorFormatter = getErrorFormatter()
+const tracer = globalTracer()
+
+type ErrorMeta = {
+  id: string
+  error?: Error
+  context: Record<string, any>
+  [id: string]: any
+}
+
+type ActivityArgs = {
+  id?: string
+  parentSpan?: Span
+}
+
+let isVerbose = false
+
+/**
+ * Reporter module.
+ * @module reporter
+ */
+class Reporter {
+  /**
+   * Strip initial indentation template function.
+   */
+  stripIndent = stripIndent
+  format = chalk
+
+  /**
+   * Toggle verbosity.
+   */
+  setVerbose(_isVerbose: boolean = true): void {
+    isVerbose = _isVerbose
+  }
+
+  /**
+   * Turn off colors in error output.
+   */
+  setNoColor(isNoColor: boolean = false): void {
+    if (isNoColor) {
+      errorFormatter.withoutColors()
+    }
+
+    // disables colors in popular terminal output coloring packages
+    //  - chalk: see https://www.npmjs.com/package/chalk#chalksupportscolor
+    //  - ansi-colors: see https://github.com/doowb/ansi-colors/blob/8024126c7115a0efb25a9a0e87bc5e29fd66831f/index.js#L5-L7
+    if (isNoColor) {
+      process.env.FORCE_COLOR = `0`
+      // chalk determines color level at import time. Before we reach this point,
+      // chalk was already imported, so we need to retroactively adjust level
+      chalk.level = 0
+    }
+  }
+
+  /**
+   * Log arguments and exit process with status 1.
+   */
+  panic(errorMeta: string | ErrorMeta | Error, error?: Error | Error[]): void {
+    const reporterError = reporter.error(errorMeta, error)
+    trackError(`GENERAL_PANIC`, { error: reporterError })
+    prematureEnd()
+    process.exit(1)
+  }
+
+  panicOnBuild(
+    errorMeta: string | ErrorMeta | Error,
```

Comment:
Could `string | ErrorMeta | Error` be a type alias, which could be reused in the other reporter types

Reply:
yeah